presto-orc/src/test/java/com/facebook/presto/orc/TestTimestampWriteAndRead.java

```diff
@@ -0,0 +1,126 @@
+            sqlTimestampOf(1_650_483_250_000_507L, SESSION, MICROSECONDS), // 2022-04-20 19:34:10.000507
+            sqlTimestampOf(60_000_000_000_123_789L, SESSION, MICROSECONDS), // 3871-04-29 10:40:00.123789
+            sqlTimestampOf(230_000_000_000_999_999L, SESSION, MICROSECONDS)); // 9258-05-30 00:53:20.999999
+
+    private static final List<SqlTimestamp> MILLISECOND_VALUES = ImmutableList.of(
+            sqlTimestampOf(0L, SESSION, MILLISECONDS), // 1970-01-01 00:00:00.000
+            sqlTimestampOf(1L, SESSION, MILLISECONDS), // 1970-01-01 00:00:00.001
+            sqlTimestampOf(999L, SESSION, MILLISECONDS), // 1970-01-01 00:00:00.999
+            sqlTimestampOf(1_000L, SESSION, MILLISECONDS), // 1970-01-01 00:00:01.000
+            sqlTimestampOf(-60_000_000_000_789L, SESSION, MILLISECONDS), // 0068-09-03 13:19:59.211
+            sqlTimestampOf(-230_000_000_999_999L, SESSION, MILLISECONDS), // -5319-08-03 22:50:00.001
+            sqlTimestampOf(1_650_483_250_507L, SESSION, MILLISECONDS), // 2022-04-20 19:34:10.507
+            sqlTimestampOf(60_000_000_000_789L, SESSION, MILLISECONDS), // 3871-04-29 10:40:00.789
+            sqlTimestampOf(230_000_000_000_999L, SESSION, MILLISECONDS)); // 9258-05-30 00:53:20.999
+
+    @Test
+    public void testMicroWriteAndRead()
+            throws Exception
+    {
+        testPrestoRoundTrip(TIMESTAMP_MICROSECONDS, MICROSECOND_VALUES, TIMESTAMP_MICROSECONDS, MICROSECOND_VALUES);
+    }
+
+    @Test
+    public void testMicroWriteAndMilliRead()
+            throws Exception
+    {
+        List<SqlTimestamp> microSecondValuesInMilli = MICROSECOND_VALUES.stream()
+                .map(microTimestamp -> new SqlTimestamp(
+                        floorDiv(microTimestamp.getMicrosUtc(), 1000),
+                        microTimestamp.getSessionTimeZoneKey().get(),
+                        TimeUnit.MILLISECONDS))
+                .collect(toList());
+
+        testPrestoRoundTrip(TIMESTAMP_MICROSECONDS, MICROSECOND_VALUES, TIMESTAMP, microSecondValuesInMilli);
+    }
+
+    @Test
+    public void testMilliWriteAndMicroRead()
+            throws Exception
+    {
+        List<SqlTimestamp> milliSecondValuesInMicro = MILLISECOND_VALUES.stream()
+                .map(milliTimestamp -> new SqlTimestamp(
+                        milliTimestamp.getMillisUtc() * 1000,
+                        milliTimestamp.getSessionTimeZoneKey().get(),
+                        MICROSECONDS))
+                .collect(toList());
+
+        testPrestoRoundTrip(TIMESTAMP, MILLISECOND_VALUES, TIMESTAMP_MICROSECONDS, milliSecondValuesInMicro);
+    }
+
+    private void testPrestoRoundTrip(Type writeType, List<SqlTimestamp> writeValues, Type readType, List<SqlTimestamp> expectedValues)
+            throws Exception
+    {
+        try (TempFile tempFile = new TempFile()) {
+            writeOrcColumnsPresto(
+                    tempFile.getFile(),
+                    DWRF,
```

Comment:
Can we also test for both DWRF and ORC encodings ?

Reply:
done